src/classes/UTIL_Describe.cls

```diff
@@ -660,6 +720,108 @@ public class UTIL_Describe {
         return mapRecordTypes.get(objectName);
     }
 
+    /*******************************************************************************************************************
+    * @description Cache for the map recordTypeInfoByDeveloperName. It is a Map with
+    * RecordTypeDevName as the key and RecordTypeInfo as the value.
+    */
+    @testVisible
+    private static Map<String, Schema.RecordTypeInfo> recordTypeInfoByDeveloperName = new Map<String, Schema.RecordTypeInfo>();
+
+    /************************************************************************************************************
+     * @description Returns a map of active, user-available recordTypeInfos for a given recordTypedevName
+     * @param objectName Name of the object to retrieve recordTypeInfos for.
+     * @param recordTypedevName The record type's developer name.
+     * @return Schema.RecordTypeInfo The RecordTypeInfo for a given recordTypedevName, object
+     */
+    public static Schema.RecordTypeInfo getRecordTypeInfoByDeveloperName(String objectName, String recordTypeDevName) {
+        if (recordTypeInfoByDeveloperName.get(recordTypedevName) == null) {
+     
+            // Get the Describe Result
+            fillMapsForObject(objectName);
+            Schema.DescribeSObjectResult obj = objectDescribes.get(objectName);
+
+            // Obtain ALL Active Record Types for the given SObjectType objectType
+            String soql = 'SELECT Id, Name, DeveloperName FROM RecordType ' + 'WHERE SObjectType = \''
+                          + String.escapeSingleQuotes(obj.getName()) + '\' ' + 'AND IsActive = TRUE';
+            List<SObject> results;
+
+            try {
+                results = Database.query(soql);
+            } catch (Exception ex) {
+                results = new List<SObject>();
+            }
+
+            // Obtain the RecordTypeInfos for this SObjectType objectType
+            Map<Id, Schema.RecordTypeInfo> recordTypeInfos = obj.getRecordTypeInfosByID();
+
+            for (SObject rt : results) {
+
+                // Check RecordType IS available for the running user
+                if (!Test.isRunningTest()) {
+                    if (recordTypeInfos.get(rt.Id).isAvailable()) {
+                        recordTypeInfoByDeveloperName.put(String.valueOf(rt.get('DeveloperName')), recordTypeInfos.get(rt.Id));
+                    }
+                // We don't want our tests to depend on the profile configuration in the org. Otherwise tests
+                // might fail just because the user running them does not have certain record types assigned.
+                } else {
+                    recordTypeInfoByDeveloperName.put(String.valueOf(rt.get('DeveloperName')), recordTypeInfos.get(rt.Id));
+                }
+            }
+        }
+
+        return recordTypeInfoByDeveloperName.get(recordTypeDevName);
```

Comment:
Please collapse this into a helper method that also populates `recordTypeInfoByName` as in `getRecordTypeInfoByName`. We can do both pieces of work at once. Use this helper method in both `getRecordTypeInfoByDeveloperName` and `getRecordTypeInfoByName`.

The end state should look similar to the following.
```
        if (recordTypeInfoByDeveloperName.get(objectName) == null) {
            populateRecordTypeInfoForSObject(objectName);
        }
        return recordTypeInfoByDeveloperName.get(objectName).get(recordTypeDevName);
```

The helper method should look similar to the following.  Note that this introduces side effect for now, which we will resolve when we make a service class to handle this kind of thing.
```
        if(recordTypeInfoByDeveloperName.get(objectName) != null) {
            return;
        }

        Map<String, Schema.RecordTypeInfo> recordTypeByDeveloperName = new Map<String, Schema.RecordTypeInfo>();
        Map<String, Schema.RecordTypeInfo> recordTypeByName = new Map<String, Schema.RecordTypeInfo>();

        // Get the Describe Result
        fillMapsForObject(objectName);
        Schema.DescribeSObjectResult obj = objectDescribes.get(objectName);

        // Obtain ALL Active Record Types for the given SObjectType objectType
        String soql = 'SELECT Id, Name, DeveloperName FROM RecordType ' + 'WHERE SObjectType = \''
                        + String.escapeSingleQuotes(obj.getName()) + '\' ' + 'AND IsActive = TRUE';
        List<SObject> results;

        try {
            results = Database.query(soql);
        } catch (Exception ex) {
            results = new List<SObject>();
        }

        // Obtain the RecordTypeInfos for this SObjectType objectType
        Map<Id, Schema.RecordTypeInfo> recordTypeInfos = obj.getRecordTypeInfosByID();

        for (SObject rt : results) {

            // Check RecordType IS available for the running user
            if (!Test.isRunningTest()) {
                if (recordTypeInfos.get(rt.Id).isAvailable()) {
                    recordTypeByDeveloperName.put(String.valueOf(rt.get('DeveloperName')), recordTypeInfos.get(rt.Id));
                    recordTypeByName.put(String.valueOf(rt.get('Name')), recordTypeInfos.get(rt.Id));
                }
            // We don't want our tests to depend on the profile configuration in the org. Otherwise tests
            // might fail just because the user running them does not have certain record types assigned.
            } else {
                recordTypeByDeveloperName.put(String.valueOf(rt.get('DeveloperName')), recordTypeInfos.get(rt.Id));
                recordTypeByName.put(String.valueOf(rt.get('Name')), recordTypeInfos.get(rt.Id));
            }
        }

        recordTypeInfoByDeveloperName.put(objectName, recordTypeByDeveloperName);
        recordTypeInfoByName.put(objectName,recordTypeByName)
```